File: bindings/c/tests/test_context.c
#include <float.h>
#include <setjmp.h>
#include <stdarg.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include <cmocka.h>

#include "cpublish.h"
#include "test_utils.h"

/*
----------------------------------------------------------------------------
  Checks
*/

static void test_cpublish_context_set_none_success(void **state) {
  CPublishStatus status;

  CPublishContext *context = cpublish_context_new();
  assert_non_null(context);

  size_t len;
  len = cpublish_context_len(context, &status);
  validate_status_ok(&status);
  assert_int_equal(len, 0);
  assert_int_equal(cpublish_context_is_empty(context, &status), true);

  cpublish_context_set_none(context, "test", &status);
  validate_status_ok(&status);

  len = cpublish_context_len(context, &status);
  assert_int_equal(len, 1);
  assert_int_equal(cpublish_context_is_empty(context, &status), false);

  const CPublishValue *value = cpublish_context_get(context, "test", &status);
  assert_non_null(value);
  validate_status_ok(&status);

  assert_int_equal(cpublish_value_type(value, &status), CPublishValueTypeNone);
  validate_status_ok(&status);

  cpublish_context_destroy(context);
}

static void test_cpublish_context_set_bool_success(void **state) {
  CPublishStatus status;

  CPublishContext *context = cpublish_context_new();
  assert_non_null(context);

  size_t len;
  len = cpublish_context_len(context, &status);
  validate_status_ok(&status);
  assert_int_equal(len, 0);
  assert_int_equal(cpublish_context_is_empty(context, &status), true);

  cpublish_context_set_bool(context, "test", true, &status);
  validate_status_ok(&status);

  len = cpublish_context_len(context, &status);
  assert_int_equal(len, 1);
  assert_int_equal(cpublish_context_is_empty(context, &status), false);

  const CPublishValue *value = cpublish_context_get(context, "test", &status);
  assert_non_null(value);
  validate_status_ok(&status);

  assert_int_equal(cpublish_value_type(value, &status),
                   CPublishValueTypeBoolean);
  validate_status_ok(&status);

  assert_true(cpublish_value_bool(value, &status));
  validate_status_ok(&status);

  cpublish_context_destroy(context);
}

static void test_cpublish_context_set_int_success(void **state) {
  CPublishStatus status;

  CPublishContext *context = cpublish_context_new();
  assert_non_null(context);

  size_t len;
  len = cpublish_context_len(context, &status);
  validate_status_ok(&status);
  assert_int_equal(len, 0);
  assert_int_equal(cpublish_context_is_empty(context, &status), true);

  cpublish_context_set_int(context, "test", 1, &status);
  validate_status_ok(&status);

  len = cpublish_context_len(context, &status);
  assert_int_equal(len, 1);
  assert_int_equal(cpublish_context_is_empty(context, &status), false);

  const CPublishValue *value = cpublish_context_get(context, "test", &status);
  assert_non_null(value);
  validate_status_ok(&status);

  assert_int_equal(cpublish_value_type(value, &status),
                   CPublishValueTypeInteger);
  validate_status_ok(&status);

  assert_int_equal(cpublish_value_int(value, &status), 1);
  validate_status_ok(&status);

  cpublish_context_destroy(context);
}

static void test_cpublish_context_set_float_success(void **state) {
  CPublishStatus status;

  CPublishContext *context = cpublish_context_new();
  assert_non_null(context);

  size_t len;
  len = cpublish_context_len(context, &status);
  validate_status_ok(&status);
  assert_int_equal(len, 0);
  assert_int_equal(cpublish_context_is_empty(context, &status), true);

  cpublish_context_set_float(context, "test", 1.0, &status);
  validate_status_ok(&status);

  len = cpublish_context_len(context, &status);
  assert_int_equal(len, 1);
  assert_int_equal(cpublish_context_is_empty(context, &status), false);

  const CPublishValue *value = cpublish_context_get(context, "test", &status);
  assert_non_null(value);
  validate_status_ok(&status);

  assert_int_equal(cpublish_value_type(value, &status), CPublishValueTypeFloat);
  validate_status_ok(&status);

  assert_float_equal(cpublish_value_float(value, &status), 1.0, DBL_EPSILON);
  validate_status_ok(&status);

  cpublish_context_destroy(context);
}

static void test_cpublish_context_set_string_success(void **state) {
  CPublishStatus status;

  CPublishContext *context = cpublish_context_new();
  assert_non_null(context);

  size_t len;
  len = cpublish_context_len(context, &status);
  validate_status_ok(&status);
  assert_int_equal(len, 0);
  assert_int_equal(cpublish_context_is_empty(context, &status), true);

  cpublish_context_set_string(context, "test", "test", &status);
  validate_status_ok(&status);

  len = cpublish_context_len(context, &status);
  assert_int_equal(len, 1);
  assert_int_equal(cpublish_context_is_empty(context, &status), false);

  const CPublishValue *value = cpublish_context_get(context, "test", &status);
  assert_non_null(value);
  validate_status_ok(&status);

  assert_int_equal(cpublish_value_type(value, &status),
                   CPublishValueTypeString);
  validate_status_ok(&status);

  CPublishString result = cpublish_value_string(value, &status);
  validate_status_ok(&status);
  assert_string_equal(result.string, "test");
  validate_status_ok(&status);

  cpublish_string_destroy(&result);

  cpublish_context_destroy(context);
}

static void test_cpublish_context_set_success(void **state) {
  CPublishStatus status;

  CPublishContext *context = cpublish_context_new();
  assert_non_null(context);

  size_t len;
  len = cpublish_context_len(context, &status);
  validate_status_ok(&status);
  assert_int_equal(len, 0);
  assert_int_equal(cpublish_context_is_empty(context, &status), true);

  CPublishValue *value = cpublish_value_new_none();
  assert_non_null(value);

  cpublish_context_set(context, "test", value, &status);
  validate_status_ok(&status);
  cpublish_value_destroy(value);

  len = cpublish_context_len(context, &status);
  assert_int_equal(len, 1);
  assert_int_equal(cpublish_context_is_empty(context, &status), false);

  const CPublishValue *value_out =
      cpublish_context_get(context, "test", &status);
  assert_non_null(value_out);
  validate_status_ok(&status);

  assert_int_equal(cpublish_value_type(value_out, &status),
                   CPublishValueTypeNone);
  validate_status_ok(&status);

  cpublish_context_destroy(context);
}

static void test_cpublish_context_clone_success(void **state) {
  CPublishStatus status;

  CPublishContext *context = cpublish_context_new();
  assert_non_null(context);

  size_t len;
  len = cpublish_context_len(context, &status);
  validate_status_ok(&status);
  assert_int_equal(len, 0);
  assert_int_equal(cpublish_context_is_empty(context, &status), true);

  cpublish_context_set_none(context, "test", &status);
  validate_status_ok(&status);

  len = cpublish_context_len(context, &status);
  assert_int_equal(len, 1);
  assert_int_equal(cpublish_context_is_empty(context, &status), false);

  CPublishContext *cloned_context = cpublish_context_clone(context, &status);

  const CPublishValue *value =
      cpublish_context_get(cloned_context, "test", &status);
  assert_non_null(value);
  validate_status_ok(&status);

  assert_int_equal(cpublish_value_type(value, &status), CPublishValueTypeNone);
  validate_status_ok(&status);

  cpublish_context_destroy(context);
  cpublish_context_destroy(cloned_context);
}

int main(void) {
  const struct CMUnitTest tests[] = {
      cmocka_unit_test(test_cpublish_context_set_none_success),
      cmocka_unit_test(test_cpublish_context_set_bool_success),
      cmocka_unit_test(test_cpublish_context_set_int_success),
      cmocka_unit_test(test_cpublish_context_set_float_success),
      cmocka_unit_test(test_cpublish_context_set_string_success),
      cmocka_unit_test(test_cpublish_context_set_success),
      cmocka_unit_test(test_cpublish_context_clone_success),
  };

  return cmocka_run_group_tests(tests, NULL, NULL);
}

File: bindings/c/tests/test_context.rs
use std::{
    ffi::{CStr, CString},
    ptr::null_mut,
};

use proptest::prelude::*;

use cpublish::*;

fn arb_value() -> impl Strategy<Value = publish::Value> {
    let leaf = prop_oneof![
        Just(publish::Value::None),
        any::<bool>().prop_map(publish::Value::Boolean),
        any::<i64>().prop_map(publish::Value::Integer),
        any::<f64>().prop_map(publish::Value::Float),
        "[^\0]*".prop_map(publish::Value::String),
    ];

    leaf.prop_recursive(8, 256, 10, |inner| {
        prop_oneof![
            prop::collection::vec(inner.clone(), 0..10).prop_map(publish::Value::Array),
            prop::collection::hash_map("[^\0]*", inner, 0..10).prop_map(publish::Value::Object),
        ]
    })
}

unsafe fn to_c_value(value: &publish::Value, status: *mut CPublishStatus) -> *mut CPublishValue {
    cpublish_status_ok(status);

    match value {
        publish::Value::None => cpublish_value_new_none(),
        publish::Value::Boolean(v) => cpublish_value_new_bool(*v),
        publish::Value::Integer(v) => cpublish_value_new_int(*v),
        publish::Value::Float(v) => cpublish_value_new_float(*v),
        publish::Value::String(v) => {
            let v = CString::new::<&str>(v.as_ref()).unwrap();
            cpublish_value_new_string(v.as_ptr())
        }
        publish::Value::Array(v) => {
            let array = cpublish_value_new_array();
            for value in v {
                let value = to_c_value(value, status);

                if !status.is_null() && (*status).status != CPublishStatusType::CPublishStatusTypeOk
                {
                    return null_mut();
                }

                cpublish_value_array_push(array, value, status);

                if !status.is_null() && (*status).status != CPublishStatusType::CPublishStatusTypeOk
                {
                    return null_mut();
                }
            }
            array
        }
        publish::Value::Object(v) => {
            let object = cpublish_value_new_object();
            for (key, value) in v {
                let key = CString::new::<&str>(key.as_ref()).unwrap();
                let value = to_c_value(value, status);

                if !status.is_null() && (*status).status != CPublishStatusType::CPublishStatusTypeOk
                {
                    return null_mut();
                }

                cpublish_value_object_insert(object, key.as_ptr(), value, status);

                if !status.is_null() && (*status).status != CPublishStatusType::CPublishStatusTypeOk
                {
                    return null_mut();
                }
            }
            object
        }
    }
}

proptest! {
    #[test]
    fn test_cpublish_context_set_success(v in prop::collection::hash_map("[^\0]*", arb_value(), 0..10)) {
        unsafe {
            let context = cpublish_context_new();
            assert!(!context.is_null());
            let mut status = CPublishStatus::new_ok();

            for (key, value) in v.iter() {
                let key = CString::new(key.as_str()).unwrap();
                let in_value = to_c_value(value, &mut status);
                cpublish_context_set(context, key.as_ptr(), in_value, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                let out_value = cpublish_context_get(context, key.as_ptr(), &mut status);
                assert!(!out_value.is_null());
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

                assert_eq!(&(*out_value).value, value);
            }

            assert_eq!(cpublish_context_len(context, &mut status), v.len());
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            if v.len() == 0 {
                assert!(cpublish_context_is_empty(context, &mut status));
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            let iter = cpublish_context_iter(context, &mut status);
            assert!(!iter.is_null());
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            while !cpublish_context_iter_is_done(iter, &mut status) {
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                let key = cpublish_context_iter_key(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                let value = cpublish_context_iter_value(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                let out_value = cpublish_context_get(context, key.string, &mut status);
                assert!(!out_value.is_null());
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

                assert_eq!(&(*out_value).value, &(*value).value);

                cpublish_context_iter_next(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            cpublish_context_iter_destroy(iter);
        }
    }

    #[test]
    fn test_cpublish_context_set_helpers_success(v in prop::collection::hash_map("[^\0]*", arb_value(), 0..10)) {
        unsafe {
            let context = cpublish_context_new();
            assert!(!context.is_null());
            let mut status = CPublishStatus::new_ok();

            for (key, value) in v.iter() {
                let key = CString::new(key.as_str()).unwrap();


                match value {
                    publish::Value::None => {
                        cpublish_context_set_none(context, key.as_ptr(), &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        let out_value = cpublish_context_get(context, key.as_ptr(), &mut status);
                        assert!(!out_value.is_null());
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        assert_eq!(&(*out_value).value, value);
                    },
                    publish::Value::Boolean(v) => {
                        cpublish_context_set_bool(context, key.as_ptr(), *v, &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        let out_value = cpublish_context_get(context, key.as_ptr(), &mut status);
                        assert!(!out_value.is_null());
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        assert_eq!(&(*out_value).value, value);
                    },
                    publish::Value::Integer(v) => {
                        cpublish_context_set_int(context, key.as_ptr(), *v, &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        let out_value = cpublish_context_get(context, key.as_ptr(), &mut status);
                        assert!(!out_value.is_null());
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        assert_eq!(&(*out_value).value, value);
                    },
                    publish::Value::Float(v) => {
                        cpublish_context_set_float(context, key.as_ptr(), *v, &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        let out_value = cpublish_context_get(context, key.as_ptr(), &mut status);
                        assert!(!out_value.is_null());
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        assert_eq!(&(*out_value).value, value);
                    },
                    publish::Value::String(v) => {
                        let c_v = CString::new(v.as_str()).unwrap();
                        cpublish_context_set_string(context, key.as_ptr(), c_v.as_ptr(), &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        let out_value = cpublish_context_get(context, key.as_ptr(), &mut status);
                        assert!(!out_value.is_null());
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        assert_eq!(&(*out_value).value, value);
                    },
                    publish::Value::Array(_) => {
                        let c_v = to_c_value(value, &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        cpublish_context_set(context, key.as_ptr(), c_v, &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        let out_value = cpublish_context_get(context, key.as_ptr(), &mut status);
                        assert!(!out_value.is_null());
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        assert_eq!(&(*out_value).value, value);
                    },
                    publish::Value::Object(_) => {
                        let c_v = to_c_value(value, &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        cpublish_context_set(context, key.as_ptr(), c_v, &mut status);
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        let out_value = cpublish_context_get(context, key.as_ptr(), &mut status);
                        assert!(!out_value.is_null());
                        assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                        assert_eq!(&(*out_value).value, value);
                    },
                };
            }

            assert_eq!(cpublish_context_len(context, &mut status), v.len());
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            if v.len() == 0 {
                assert!(cpublish_context_is_empty(context, &mut status));
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }
        }
    }

    #[test]
    fn test_cpublish_context_clone_success(v in prop::collection::hash_map("[^\0]*", arb_value(), 0..10)) {
        unsafe {
            let context = cpublish_context_new();
            assert!(!context.is_null());
            let mut status = CPublishStatus::new_ok();

            for (key, value) in v.iter() {
                let key = CString::new(key.as_str()).unwrap();
                let in_value = to_c_value(value, &mut status);
                cpublish_context_set(context, key.as_ptr(), in_value, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            let cloned_context = cpublish_context_clone(context, &mut status);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            let context_len = cpublish_context_len(context, &mut status);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            let cloned_context_len = cpublish_context_len(cloned_context, &mut status);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            assert_eq!(context_len, cloned_context_len);

            let iter = cpublish_context_iter(context, &mut status);

            while !cpublish_context_iter_is_done(iter, &mut status) {
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                let key = cpublish_context_iter_key(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

                let value = cpublish_context_iter_value(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                let cloned_value = cpublish_context_get(cloned_context, key.string, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

                assert_eq!(&(*value).value, &(*cloned_value).value);

                cpublish_context_iter_next(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            cpublish_context_iter_destroy(iter);
        }
    }
}
